test-fixture/src/sim/taildrop.rs

```diff
@@ -124,38 +213,92 @@ impl TailDrop {
     }
 
     /// Enqueue for sending.  Maybe.  If this overflows the queue, drop it instead.
+    /// If the queue is too deep, CE mark it before saving.
     fn maybe_enqueue(&mut self, d: Datagram, now: Instant) {
-        self.received += 1;
+        self.stats.received += 1;
         if self.next_deque.is_none() {
             // Nothing in the queue and nothing still sending.
             debug_assert!(self.queue.is_empty());
-            self.send(d, now);
+            self.send(d, now, 0);
         } else if self.used + self.size(&d) <= self.capacity {
+            self.stats.accumulate_usage(now, self.used);
             self.used += self.size(&d);
-            self.maxq = max(self.maxq, self.used);
+            self.stats.update_maxq(self.used);
+            let d = self.maybe_mark(d);
             self.queue.push_back(d);
         } else {
             qtrace!("taildrop dropping {} bytes", d.len());
-            self.dropped += 1;
+            self.stats.dropped += 1;
         }
     }
 
     /// If the last packet that was sending has been sent, start sending
     /// the next one.
     fn maybe_send(&mut self, now: Instant) {
-        if self.next_deque.as_ref().is_some_and(|t| *t <= now) {
+        if let Some((t, subns)) = &self.next_deque {
+            if now < *t {
+                return;
+            }
+
+            self.stats.accumulate_usage(now, self.used);
             if let Some(d) = self.queue.pop_front() {
                 self.used -= self.size(&d);
-                self.send(d, now);
+                self.send(d, *t, *subns);
             } else {
                 self.next_deque = None;
-                self.sub_ns_delay = 0;
             }
         }
     }
+
+    /// Apply ECN-CE markings to packets if they are ECT(0) marked.
+    /// This is classic, simple ECN using RED; we don't know about L4S yet.
+    fn maybe_mark(&mut self, dgram: Datagram) -> Datagram {
+        if self.ecn && Ecn::from(dgram.tos()) == Ecn::Ect0 && self.should_mark(self.used) {
+            qtrace!("taildrop marking {} bytes", dgram.len());
+            self.stats.marked += 1;
+            let tos = Tos::from((Dscp::from(dgram.tos()), Ecn::Ce));
+            Datagram::new(dgram.source(), dgram.destination(), tos, dgram.to_vec())
+        } else {
+            dgram
+        }
+    }
```

Comment:
Also apply CE for ECT(1), they should be identical. (Not that we use ECT(1) yet.)

Reply:
OK, good.  I was under the impression that we'd need a different algorithm.  Easy fix.